File: app/src/main/java/in/engineerakash/covid19india/ui/home/DistrictWiseAdapter.java
package in.engineerakash.covid19india.ui.home;

import android.content.Context;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import in.engineerakash.covid19india.R;
import in.engineerakash.covid19india.databinding.DistrictDataItemBinding;
import in.engineerakash.covid19india.pojo.District;
import in.engineerakash.covid19india.util.Constant;

public class DistrictWiseAdapter extends RecyclerView.Adapter<DistrictWiseAdapter.DistrictWiseVH> {

    private ArrayList<District> list;

    private int TYPE_ITEM = 1;
    private int TYPE_EMPTY_VIEW = 2;
    private Context context;

    public DistrictWiseAdapter(ArrayList<District> list) {
        this.list = list;
    }

    @NonNull
    @Override
    public DistrictWiseVH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        context = parent.getContext();

        DistrictDataItemBinding binding = DistrictDataItemBinding.inflate(LayoutInflater.from(context),
                parent, false);

        return new DistrictWiseVH(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull DistrictWiseVH holder, int position) {

        /*if (position == 0) {
            // show header
            holder.binding.districtNameTv.setText("District");
            holder.binding.districtConfirmedTv.setText("Confirmed");
            holder.binding.districtLastUpdatedTv.setText("Last updated at");

            holder.binding.userDistrictTv.setVisibility(View.GONE);

            holder.binding.dataContainer.setVisibility(View.VISIBLE);
            holder.binding.emptyView.setVisibility(View.GONE);

        } else*/
        if (getItemViewType(position) == TYPE_EMPTY_VIEW) {
            // show empty view

            holder.binding.dataContainer.setVisibility(View.GONE);
            holder.binding.emptyView.setVisibility(View.VISIBLE);

            holder.binding.getRoot().setBackgroundColor(ContextCompat.getColor(context, R.color.colorOddItem));

        } else {

            District data = list.get(position);

            // show data
            holder.binding.districtNameTv.setText(data.getName());
            holder.binding.districtConfirmedTv.setText(String.valueOf(data.getConfirmed()));
            holder.binding.districtLastUpdatedTv.setText(data.getLastUpdateTime().isEmpty() ? "-" : data.getLastUpdateTime());

            if (Constant.userSelectedDistrict.trim().equalsIgnoreCase(data.getName().trim()) &&
                    Constant.isUserSelected)
                holder.binding.userDistrictTv.setVisibility(View.VISIBLE);
            else
                holder.binding.userDistrictTv.setVisibility(View.GONE);

            holder.binding.dataContainer.setVisibility(View.VISIBLE);
            holder.binding.emptyView.setVisibility(View.GONE);

            if (position % 2 == 0) {
                holder.binding.getRoot().setBackgroundColor(ContextCompat.getColor(context, R.color.colorOddItem));
            } else {
                holder.binding.getRoot().setBackgroundColor(ContextCompat.getColor(context, R.color.colorEvenItem));
            }

            // Total element
            if (position == getItemCount() - 1) {
                holder.binding.districtNameTv.setTypeface(holder.binding.districtNameTv.getTypeface(), Typeface.BOLD);
                holder.binding.districtConfirmedTv.setTypeface(holder.binding.districtConfirmedTv.getTypeface(), Typeface.BOLD);
                holder.binding.districtLastUpdatedTv.setTypeface(holder.binding.districtLastUpdatedTv.getTypeface(), Typeface.BOLD);
            } else {
                holder.binding.districtNameTv.setTypeface(holder.binding.districtNameTv.getTypeface(), Typeface.NORMAL);
                holder.binding.districtConfirmedTv.setTypeface(holder.binding.districtConfirmedTv.getTypeface(), Typeface.NORMAL);
                holder.binding.districtLastUpdatedTv.setTypeface(holder.binding.districtLastUpdatedTv.getTypeface(), Typeface.NORMAL);
            }
        }

    }

    @Override
    public int getItemCount() {
        if (list.isEmpty())
            return 1; // empty view
        else
            return list.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (list.isEmpty())
            return TYPE_EMPTY_VIEW;
        else
            return TYPE_ITEM;
    }

    static class DistrictWiseVH extends RecyclerView.ViewHolder {
        private DistrictDataItemBinding binding;

        public DistrictWiseVH(@NonNull DistrictDataItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/in/engineerakash/covid19india/ui/track/DistrictWiseAdapter.kt
package `in`.engineerakash.covid19india.ui.track

import `in`.engineerakash.covid19india.R
import `in`.engineerakash.covid19india.databinding.DistrictDataItemBinding
import `in`.engineerakash.covid19india.pojo.District
import `in`.engineerakash.covid19india.util.Constant
import `in`.engineerakash.covid19india.util.ViewUtil.isNullOr0
import android.content.Context
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import java.util.*

private const val TYPE_ITEM = 1
private const val TYPE_EMPTY_VIEW = 2

class DistrictWiseAdapter(var list: ArrayList<District>, var allowScrollAnimation: Boolean = true) :
    RecyclerView.Adapter<DistrictWiseAdapter.DistrictWiseVH>() {

    private var context: Context? = null
    private var lastPosition = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DistrictWiseVH {
        context = parent.context
        val binding = DistrictDataItemBinding.inflate(
            LayoutInflater.from(context), parent, false
        )
        return DistrictWiseVH(binding)
    }


    override fun onBindViewHolder(holder: DistrictWiseVH, position: Int) {
        holder.bind(position)
    }

    override fun onViewDetachedFromWindow(holder: DistrictWiseVH) {
        super.onViewDetachedFromWindow(holder)
        if (allowScrollAnimation)
            holder.itemView.clearAnimation()
    }


    override fun getItemCount(): Int {
        return if (list.isEmpty()) 1 // empty view
        else list.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (list.isEmpty()) TYPE_EMPTY_VIEW else TYPE_ITEM
    }

    inner class DistrictWiseVH(private val itemBinding: DistrictDataItemBinding) :
        RecyclerView.ViewHolder(itemBinding.root) {

        fun bind(position: Int) {

            if (getItemViewType(position) == TYPE_EMPTY_VIEW) {
                // show empty view
                itemBinding.dataContainer.visibility = View.GONE
                itemBinding.emptyView.visibility = View.VISIBLE
                itemBinding.root.setBackgroundColor(
                    ContextCompat.getColor(itemBinding.root.context, R.color.colorOddItem)
                )
            } else {
                val data: District = list[position]

                // show data
                itemBinding.districtNameTv.text = data.name

                itemBinding.districtConfirmedTv.text = (data.total?.confirmed ?: "-").toString()
                itemBinding.deltaDistrictConfirmedTv.text =
                    (data.delta?.confirmed ?: "-").toString()

                itemBinding.districtRecoveredTv.text = (data.total?.recovered ?: "-").toString()
                itemBinding.deltaDistrictRecoveredTv.text =
                    (data.delta?.recovered ?: "-").toString()

                itemBinding.districtDeathTv.text = (data.total?.deceased ?: "-").toString()
                itemBinding.deltaDistrictDeathTv.text = (data.delta?.deceased ?: "-").toString()

                if (Constant.userSelectedDistrict.trim { it <= ' ' }
                        .equals(data.name.trim { it <= ' ' }, ignoreCase = true) &&
                    Constant.locationIsSelectedByUser
                )
                    itemBinding.userDistrictTv.visibility = View.VISIBLE
                else
                    itemBinding.userDistrictTv.visibility = View.GONE
                itemBinding.dataContainer.visibility = View.VISIBLE
                itemBinding.emptyView.visibility = View.GONE
                if (position % 2 == 0) {
                    itemBinding.root.setBackgroundColor(
                        ContextCompat.getColor(itemBinding.root.context, R.color.colorOddItem)
                    )
                } else {
                    itemBinding.root.setBackgroundColor(
                        ContextCompat.getColor(itemBinding.root.context, R.color.colorEvenItem)
                    )
                }

                // Total element
                if (data.name.equals(context?.getString(R.string.total), ignoreCase = true)) {
                    itemBinding.districtNameTv.typeface = Typeface.defaultFromStyle(Typeface.BOLD)
                    itemBinding.districtConfirmedTv.typeface =
                        Typeface.defaultFromStyle(Typeface.BOLD)
                    itemBinding.districtRecoveredTv.typeface =
                        Typeface.defaultFromStyle(Typeface.BOLD)
                    itemBinding.districtDeathTv.typeface = Typeface.defaultFromStyle(Typeface.BOLD)
                } else {
                    itemBinding.districtNameTv.typeface = Typeface.defaultFromStyle(Typeface.NORMAL)
                    itemBinding.districtConfirmedTv.typeface =
                        Typeface.defaultFromStyle(Typeface.NORMAL)
                    itemBinding.districtRecoveredTv.typeface =
                        Typeface.defaultFromStyle(Typeface.NORMAL)
                    itemBinding.districtDeathTv.typeface =
                        Typeface.defaultFromStyle(Typeface.NORMAL)
                }

                if (data.delta?.confirmed.isNullOr0() && data.delta?.recovered.isNullOr0() && data.delta?.deceased.isNullOr0()) {
                    itemBinding.deltaDistrictConfirmedTv.visibility = View.GONE
                    itemBinding.deltaDistrictRecoveredTv.visibility = View.GONE
                    itemBinding.deltaDistrictDeathTv.visibility = View.GONE
                } else {

                    if (data.delta?.confirmed.isNullOr0())
                        itemBinding.deltaDistrictConfirmedTv.visibility = View.INVISIBLE
                    else
                        itemBinding.deltaDistrictConfirmedTv.visibility = View.VISIBLE

                    if (data.delta?.recovered.isNullOr0())
                        itemBinding.deltaDistrictRecoveredTv.visibility = View.INVISIBLE
                    else
                        itemBinding.deltaDistrictRecoveredTv.visibility = View.VISIBLE

                    if (data.delta?.deceased.isNullOr0())
                        itemBinding.deltaDistrictDeathTv.visibility = View.INVISIBLE
                    else
                        itemBinding.deltaDistrictDeathTv.visibility = View.VISIBLE
                }
            }

            if (allowScrollAnimation) {
                val animation: Animation = AnimationUtils.loadAnimation(
                    context,
                    if (position > lastPosition) R.anim.up_from_bottom else R.anim.down_from_top
                )
                itemView.startAnimation(animation)
                lastPosition = position
            }

        }
    }
}